import pandas as pd
import os
# ^^^ pyforest auto-imports - don't write above this line
# Clinical Deterioration Prediction Model - Preprocessing II


## Data Source 

For this project, I used publicly available Electronic Health Records (EHRs) datasets. The MIT Media Lab for Computational Physiology has developed MIMIC-IIIv1.4 dataset based on 46,520 patients who stayed in critical care units of the Beth Israel Deaconess Medical Center of Boston between 2001 and 2012. MIMIC-IIIv1.4 dataset is freely available to researchers across the world. A formal request should be made directly to www.mimic.physionet.org, to gain acess to the data. There is a required course on human research ‘Data or Specimens Only Research’ prior to data acess request. I have secured one here -www.citiprogram.org/verify/?kb6607b78-5821-4de5-8cad-daf929f7fbbf-33486907

The dataset has 26 relational tables including patient’s hospital admission, callout information when patient was ready for discharge, caregiver information, electronic charted events including vital signs and any additional information relevant to patient care, patient demographic data, list of services the patient was admitted or transferred under,  ICU stay types, diagnoses types, laboratory measurments, microbiology tests and sensitivity, prescription data and billing information. 

Although I have full access to the MIMIC-IIIv1.4 datasets, I can not share any part of the data publicly. If you are interested to learn more about the data, there is a MIMIC III Demo dataset based on 100 patients https://mimic.physionet.org/gettingstarted/demo/. If you are interested to requesting access to the data - https://mimic.physionet.org/gettingstarted/access/.

In [1]:
import os
import pandas as pd
import numpy as np
import random
import sys
import csv

In [2]:
os.getcwd()

<IPython.core.display.Javascript object>

'C:\\Users\\abebu\\Dropbox\\Data Science\\Projects\\Capstone Project 1\\Potential Projects\\9. MIMIC\\data_wrangling'

In [3]:
os.chdir("C://Users/abebu/Google Drive/mimic-iii-clinical-database-1.4")

<IPython.core.display.Javascript object>

In [4]:
saps=pd.read_csv('charts_f.csv', header=0, index_col=0)
saps

C:\Users\abebu\DS\lib\site-packages\IPython\core\interactiveshell.py:3058: DtypeWarning: Columns (18,19,21,22,23,25,26,27,28) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)
C:\Users\abebu\DS\lib\site-packages\numpy\lib\arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


SUBJECT_ID  HADM_ID  ICUSTAY_ID  ART BP Systolic   BUN  \
0                  2   163353    243653.0              NaN   NaN   
1                  2   163353    243653.0              NaN   NaN   
2                  2   163353    243653.0              NaN   NaN   
3                  2   163353    243653.0              NaN   NaN   
4                  2   163353    243653.0             72.0   NaN   
...              ...      ...         ...              ...   ...   
25100118       99999   113369    246512.0              NaN   NaN   
25100119       99999   113369    246512.0              NaN   NaN   
25100120       99999   113369    246512.0              NaN   NaN   
25100121       99999   113369    246512.0              NaN  15.0   
25100122       99999   113369    246512.0              NaN   NaN   

          GCS - Eye Opening  GCS - Motor Response  GCS - Verbal Response  \
0                       NaN                   NaN                    NaN   
1                       NaN                   NaN                    NaN   
2                       NaN                   NaN                    NaN   
3                       NaN                   NaN                    NaN   
4                       NaN                   NaN                    NaN   
...                     ...                   ...                    ...   
25100118                NaN                   6.0                    NaN   
25100119                NaN                   6.0                    NaN   
25100120                NaN                   6.0                    NaN   
25100121                NaN                   NaN                    NaN   
25100122                NaN                   NaN                    NaN   

          GCS Total  HCO3 (serum)  ...   ADMITTIME            DISCHTIME  \
0               NaN           NaN  ...  2138-07-17  2138-07-21 15:48:00   
1               NaN           NaN  ...  2138-07-17  2138-07-21 15:48:00   
2               NaN           NaN  ...  2138-07-17  2138-07-21 15:48:00   
3               NaN           NaN  ...  2138-07-17  2138-07-21 15:48:00   
4               NaN           NaN  ...  2138-07-17  2138-07-21 15:48:00   
...             ...           ...  ...         ...                  ...   
25100118        NaN           NaN  ...  2117-12-30  2118-01-04 16:30:00   
25100119        NaN           NaN  ...  2117-12-30  2118-01-04 16:30:00   
25100120        NaN           NaN  ...  2117-12-30  2118-01-04 16:30:00   
25100121        NaN           NaN  ...  2117-12-30  2118-01-04 16:30:00   
25100122        NaN          26.0  ...  2117-12-30  2118-01-04 16:30:00   

          ADMISSION_TYPE  HOSPITAL_EXPIRE_FLAG  GENDER         DOB  DOD  \
0                NEWBORN                   0.0       M  2138-07-17  NaN   
1                NEWBORN                   0.0       M  2138-07-17  NaN   
2                NEWBORN                   0.0       M  2138-07-17  NaN   
3                NEWBORN                   0.0       M  2138-07-17  NaN   
4                NEWBORN                   0.0       M  2138-07-17  NaN   
...                  ...                   ...     ...         ...  ...   
25100118        ELECTIVE                   0.0       F  2054-01-06  NaN   
25100119        ELECTIVE                   0.0       F  2054-01-06  NaN   
25100120        ELECTIVE                   0.0       F  2054-01-06  NaN   
25100121        ELECTIVE                   0.0       F  2054-01-06  NaN   
25100122        ELECTIVE                   0.0       F  2054-01-06  NaN   

         DOD_HOSP EXPIRE_FLAG        AGE  
0             NaN         0.0   0.000000  
1             NaN         0.0   0.000000  
2             NaN         0.0   0.000000  
3             NaN         0.0   0.000000  
4             NaN         0.0   0.000000  
...           ...         ...        ...  
25100118      NaN         0.0  64.021918  
25100119      NaN         0.0  64.021918  
25100120      NaN         0.0  64.021918  
25100121      NaN         0.0  64.021918  
25100

In [5]:
saps=saps.drop(['DOD', 'DOD_HOSP', 'DOB', 'ADMITTIME', 'DISCHTIME'], axis=1)

In [7]:
saps=saps.drop(['INTIME', 'OUTTIME'], axis=1)

In [8]:
saps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 25100123 entries, 0 to 25100122
Data columns (total 23 columns):
SUBJECT_ID                 int64
HADM_ID                    int64
ICUSTAY_ID                 float64
ART BP Systolic            float64
BUN                        float64
GCS - Eye Opening          float64
GCS - Motor Response       float64
GCS - Verbal Response      float64
GCS Total                  float64
HCO3 (serum)               float64
Heart Rate                 float64
Inspired O2 Fraction       float64
Potassium (whole blood)    float64
Sodium (serum)             float64
Temperature Celsius        float64
Total Bilirubin            float64
WBC                        float64
LOS                        float64
ADMISSION_TYPE             object
HOSPITAL_EXPIRE_FLAG       float64
GENDER                     object
EXPIRE_FLAG                float64
AGE                        float64
dtypes: float64(19), int64(2), object(2)
memory usage: 4.5+ GB


In [13]:
saps

BUN  GCS - Eye Opening  ART BP Systolic  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                             
2          163353  243653.0     NaN                NaN             72.0   
3          145834  211552.0    17.0                NaN              0.0   
4          185777  294638.0    10.0                NaN             97.0   
5          178980  214757.0     NaN                NaN             67.0   
6          107064  228232.0    65.0                NaN            125.0   
...                             ...                ...              ...   
99985      176670  279638.0    13.0                2.0             71.0   
99991      151118  226241.0    17.0                3.0            112.0   
99992      197084  242052.0    44.0                3.0             83.0   
99995      137810  229633.0    10.0                3.0            104.0   
99999      113369  246512.0    15.0                4.0            100.0   

                               GCS - Motor Response  GCS - Verbal Response  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                                
2          163353  243653.0                     NaN                    NaN   
3          145834  211552.0                     NaN                    NaN   
4          185777  294638.0                     NaN                    NaN   
5          178980  214757.0                     NaN                    NaN   
6          107064  228232.0                     NaN                    NaN   
...                                             ...                    ...   
99985      176670  279638.0                     4.0                    1.0   
99991      151118  226241.0                     6.0                    5.0   
99992      197084  242052.0                     6.0                    5.0   
99995      137810  229633.0                     6.0                    1.0   
99999      113369  246512.0                     6.0                    5.0   

                               GCS Total  HCO3 (serum)  Heart Rate  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                        
2          163353  243653.0          NaN           NaN       131.0   
3          145834  211552.0          3.0           NaN        64.0   
4          185777  294638.0         15.0           NaN        74.0   
5          178980  214757.0          NaN           NaN       140.0   
6          107064  228232.0         14.0           NaN        71.0   
...                                  ...           ...         ...   
99985      176670  279638.0          NaN          21.0         5.0   
99991      151118  226241.0          NaN          21.0        50.0   
99992      197084  242052.0          NaN          22.0        50.0   
99995      137810  229633.0          NaN          32.0        45.0   
99999      113369  246512.0          NaN          26.0        50.0   

                               Inspired O2 Fraction  Potassium (whole blood)  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                                  
2          163353  243653.0                     NaN                      NaN   
3          145834  211552.0                    20.0                      3.4   
4          185777  294638.0                     NaN                      3.3   
5          178980  214757.0                     NaN                      NaN   
6          107064  228232.0                     NaN                      4.4   
...                                             ...                      ...   
99985      176670  279638.0                    39.0                      NaN   
99991      151118  226241.0                     0.0                      NaN   
99992      197084  242052.0                     NaN                      NaN   
99995      137810  229633.0                    50.0                      3.2   
99999      113369  246512.0                    35.0                      NaN   

                               Sodium (serum)  Temperature Celsius  \
SUBJECT_ID HADM_ID 

In [18]:
  
saps.loc[saps['Heart Rate'] <40, 'HR'] = 11
saps.loc[saps['Heart Rate'].between(40, 59), 'HR'] = 2
saps.loc[saps['Heart Rate'].between(70, 119), 'HR'] = 0
saps.loc[saps['Heart Rate'].between(120, 159), 'HR'] = 4
saps.loc[saps['Heart Rate'] >159, 'HR'] = 7

BUN  GCS - Eye Opening  ART BP Systolic  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                             
2          163353  243653.0     NaN                NaN             72.0   
3          145834  211552.0    17.0                NaN              0.0   
4          185777  294638.0    10.0                NaN             97.0   
5          178980  214757.0     NaN                NaN             67.0   
6          107064  228232.0    65.0                NaN            125.0   
...                             ...                ...              ...   
99985      176670  279638.0    13.0                2.0             71.0   
99991      151118  226241.0    17.0                3.0            112.0   
99992      197084  242052.0    44.0                3.0             83.0   
99995      137810  229633.0    10.0                3.0            104.0   
99999      113369  246512.0    15.0                4.0            100.0   

                               GCS - Motor Response  GCS - Verbal Response  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                                
2          163353  243653.0                     NaN                    NaN   
3          145834  211552.0                     NaN                    NaN   
4          185777  294638.0                     NaN                    NaN   
5          178980  214757.0                     NaN                    NaN   
6          107064  228232.0                     NaN                    NaN   
...                                             ...                    ...   
99985      176670  279638.0                     4.0                    1.0   
99991      151118  226241.0                     6.0                    5.0   
99992      197084  242052.0                     6.0                    5.0   
99995      137810  229633.0                     6.0                    1.0   
99999      113369  246512.0                     6.0                    5.0   

                               GCS Total  HCO3 (serum)  Heart Rate  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                        
2          163353  243653.0          NaN           NaN       131.0   
3          145834  211552.0          3.0           NaN        64.0   
4          185777  294638.0         15.0           NaN        74.0   
5          178980  214757.0          NaN           NaN       140.0   
6          107064  228232.0         14.0           NaN        71.0   
...                                  ...           ...         ...   
99985      176670  279638.0          NaN          21.0         5.0   
99991      151118  226241.0          NaN          21.0        50.0   
99992      197084  242052.0          NaN          22.0        50.0   
99995      137810  229633.0          NaN          32.0        45.0   
99999      113369  246512.0          NaN          26.0        50.0   

                               Inspired O2 Fraction  Potassium (whole blood)  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                                  
2          163353  243653.0                     NaN                      NaN   
3          145834  211552.0                    20.0                      3.4   
4          185777  294638.0                     NaN                      3.3   
5          178980  214757.0                     NaN                      NaN   
6          107064  228232.0                     NaN                      4.4   
...                                             ...                      ...   
99985      176670  279638.0                    39.0                      NaN   
99991      151118  226241.0                     0.0                      NaN   
99992      197084  242052.0                     NaN                      NaN   
99995      137810  229633.0                    50.0                      3.2   
99999      113369  246512.0                    35.0                      NaN   

                               Sodium (serum)  Temperature Celsius  \
SUBJECT_ID HADM_ID 

In [19]:
saps.loc[saps['BUN'] <28, 'BUN'] = 0
saps.loc[saps['BUN'].between(28, 83), 'BUN'] = 6
saps.loc[saps['BUN'] >83, 'BUN'] = 10

In [20]:
saps['GCS_SCORE']=saps['GCS - Eye Opening'] + saps['GCS - Motor Response'] + saps['GCS - Verbal Response']

In [22]:
saps=saps.drop(['GCS - Eye Opening', 'GCS - Motor Response', 'GCS - Verbal Response'], axis=1)

In [24]:
saps.GCS_SCORE.fillna(saps['GCS Total'], inplace=True)

In [26]:
del saps['GCS Total']

In [28]:
saps.loc[saps['GCS_SCORE'] <6, 'GCS_SCORE'] = 26
saps.loc[saps['GCS_SCORE'].between(6, 8), 'GCS_SCORE'] = 13
saps.loc[saps['GCS_SCORE'].between(9, 10), 'GCS_SCORE'] = 7
saps.loc[saps['GCS_SCORE'].between(11, 13), 'GCS_SCORE'] = 5
saps.loc[saps['GCS_SCORE'].between(14, 15), 'GCS_SCORE'] = 0

In [29]:
saps.loc[saps['Temperature Celsius'] <39, 'Temperature Celsius'] = 3
saps.loc[saps['Temperature Celsius'] >38.9, 'Temperature Celsius'] = 0


In [31]:
saps.loc[saps['ART BP Systolic'] <70, 'ART BP Systolic'] = 13
saps.loc[saps['ART BP Systolic'].between(70, 99), 'ART BP Systolic'] = 5
saps.loc[saps['ART BP Systolic'].between(100, 199), 'ART BP Systolic'] = 0
saps.loc[saps['ART BP Systolic']>199, 'ART BP Systolic'] = 2

In [32]:
saps.loc[saps['HCO3 (serum)'] <15, 'HCO3 (serum)'] = 6
saps.loc[saps['HCO3 (serum)'].between(15, 19), 'HCO3 (serum)'] = 3
saps.loc[saps['HCO3 (serum)']>19, 'HCO3 (serum)'] = 0

In [33]:
saps.loc[saps['Total Bilirubin'] <4, 'Total Bilirubin'] = 0
saps.loc[saps['Total Bilirubin'].between(4, 5.9), 'Total Bilirubin'] = 4
saps.loc[saps['Total Bilirubin']>5.9, 'Total Bilirubin'] = 9

In [34]:
saps.loc[saps['Potassium (whole blood)'] <3, 'Potassium (whole blood)'] = 3
saps.loc[saps['Potassium (whole blood)'].between(3, 4.9), 'Potassium (whole blood)'] = 0
saps.loc[saps['Potassium (whole blood)']>4.9, 'Potassium (whole blood)'] = 3

In [35]:
saps.loc[saps['Sodium (serum)'] <125, 'Sodium (serum)'] = 5
saps.loc[saps['Sodium (serum)'].between(125, 144), 'Sodium (serum)'] = 0
saps.loc[saps['Sodium (serum)']>144, 'Sodium (serum)'] = 1

In [36]:
saps.loc[saps['WBC'] <1, 'WBC'] = 12
saps.loc[saps['WBC'].between(1, 19.9), 'WBC'] = 0
saps.loc[saps['WBC']>19.9, 'WBC'] = 3

In [37]:
saps.loc[saps['Inspired O2 Fraction'] <13.3, 'Inspired O2 Fraction'] = 11
saps.loc[saps['Inspired O2 Fraction'].between(13.3, 26.5), 'Inspired O2 Fraction'] = 9
saps.loc[saps['Inspired O2 Fraction']>26.5, 'Inspired O2 Fraction'] = 6
saps['Inspired O2 Fraction'] = saps['Inspired O2 Fraction'].fillna(0)

In [42]:
#ART BP Systolic
del saps['Heart Rate']

In [45]:
#rename columns
saps = saps.rename(columns = {'HCO3 (serum)':'Bicarbonate','Inspired O2 Fraction':'ventilation', 
                              'Potassium (whole blood)': 'Potassium', 'Sodium (serum)':'Sodium', 
                             'Temperature Celsius': 'Temp', 'Total Bilirubin':'Bilirubin',
                             'GCS_SCORE':'GCS'})

In [46]:
saps

BUN  Bicarbonate  ventilation  Potassium  \
SUBJECT_ID HADM_ID ICUSTAY_ID                                             
2          163353  243653.0    NaN          NaN          0.0        NaN   
3          145834  211552.0    0.0          NaN          9.0        0.0   
4          185777  294638.0    0.0          NaN          0.0        0.0   
5          178980  214757.0    NaN          NaN          0.0        NaN   
6          107064  228232.0    6.0          NaN          0.0        0.0   
...                            ...          ...          ...        ...   
99985      176670  279638.0    0.0          0.0          6.0        NaN   
99991      151118  226241.0    0.0          0.0         11.0        NaN   
99992      197084  242052.0    6.0          0.0          0.0        NaN   
99995      137810  229633.0    0.0          0.0          6.0        0.0   
99999      113369  246512.0    0.0          0.0          6.0        NaN   

                               Sodium  Temp  Bilirubin  WBC    HR   GCS  
SUBJECT_ID HADM_ID ICUSTAY_ID                                            
2          163353  243653.0       NaN   NaN        NaN  NaN   4.0   NaN  
3          145834  211552.0       0.0   3.0        NaN  0.0   NaN  26.0  
4          185777  294638.0       0.0   3.0        0.0  0.0   0.0   0.0  
5          178980  214757.0       NaN   NaN        NaN  0.0   4.0   NaN  
6          107064  228232.0       0.0   3.0        0.0  0.0   0.0   0.0  
...                               ...   ...        ...  ...   ...   ...  
99985      176670  279638.0       0.0   0.0        0.0  0.0  11.0   5.0  
99991      151118  226241.0       0.0   3.0        NaN  0.0   2.0   0.0  
99992      197084  242052.0       0.0   3.0        NaN  0.0   2.0   0.0  
99995      137810  229633.0       0.0   3.0        0.0  0.0   2.0   7.0  
99999      113369  246512.0       0.0   3.0        NaN  0.0   2.0   0.0  

[60661 rows x 10 columns]

In [47]:
fields = ['ICUSTAY_ID', 'ART BP Systolic']
bp=pd.read_csv('charts_f.csv', usecols=fields, header=0, index_col=0)
bp

C:\Users\abebu\DS\lib\site-packages\numpy\lib\arraysetops.py:569: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  mask |= (ar1 == a)


ART BP Systolic
ICUSTAY_ID                 
243653.0                NaN
243653.0                NaN
243653.0                NaN
243653.0                NaN
243653.0               72.0
...                     ...
246512.0                NaN
246512.0                NaN
246512.0                NaN
246512.0                NaN
246512.0                NaN

[25100123 rows x 1 columns]

In [49]:
bp.loc[bp['ART BP Systolic'] <70, 'ART BP Systolic'] = 13
bp.loc[bp['ART BP Systolic'].between(70, 99), 'ART BP Systolic'] = 5
bp.loc[bp['ART BP Systolic'].between(100, 199), 'ART BP Systolic'] = 0
bp.loc[bp['ART BP Systolic']>199, 'ART BP Systolic'] = 2

In [51]:
bp=bp.groupby('ICUSTAY_ID').agg('max')

In [53]:
bp= bp.rename(columns = {'ART BP Systolic':'SystolicBP'})

In [54]:
bp

SystolicBP
ICUSTAY_ID            
200001.0           5.0
200003.0          13.0
200006.0           5.0
200007.0           5.0
200009.0           5.0
...                ...
299993.0           0.0
299994.0          13.0
299995.0           0.0
299998.0           5.0
299999.0           5.0

[60517 rows x 1 columns]

In [55]:
def icu_merge(table1, table2):
    return table1.merge(table2, how='left', left_on=['ICUSTAY_ID'], right_on=['ICUSTAY_ID'])

In [56]:
saps=icu_merge(saps, bp)
saps

BUN  Bicarbonate  ventilation  Potassium  Sodium  Temp  Bilirubin  \
ICUSTAY_ID                                                                      
243653.0    NaN          NaN          0.0        NaN     NaN   NaN        NaN   
211552.0    0.0          NaN          9.0        0.0     0.0   3.0        NaN   
294638.0    0.0          NaN          0.0        0.0     0.0   3.0        0.0   
214757.0    NaN          NaN          0.0        NaN     NaN   NaN        NaN   
228232.0    6.0          NaN          0.0        0.0     0.0   3.0        0.0   
...         ...          ...          ...        ...     ...   ...        ...   
279638.0    0.0          0.0          6.0        NaN     0.0   0.0        0.0   
226241.0    0.0          0.0         11.0        NaN     0.0   3.0        NaN   
242052.0    6.0          0.0          0.0        NaN     0.0   3.0        NaN   
229633.0    0.0          0.0          6.0        0.0     0.0   3.0        0.0   
246512.0    0.0          0.0          6.0        NaN     0.0   3.0        NaN   

            WBC    HR   GCS  SystolicBP  
ICUSTAY_ID                               
243653.0    NaN   4.0   NaN         5.0  
211552.0    0.0   NaN  26.0        13.0  
294638.0    0.0   0.0   0.0         5.0  
214757.0    0.0   4.0   NaN        13.0  
228232.0    0.0   0.0   0.0         0.0  
...         ...   ...   ...         ...  
279638.0    0.0  11.0   5.0         5.0  
226241.0    0.0   2.0   0.0         2.0  
242052.0    0.0   2.0   0.0         5.0  
229633.0    0.0   2.0   7.0         0.0  
246512.0    0.0   2.0   0.0         0.0  

[60661 rows x 11 columns]

In [5]:
fields = ['ICUSTAY_ID', 'UO']
uo=pd.read_csv('uo_value.csv', usecols=fields, header=0, index_col=0)
uo

<IPython.core.display.Javascript object>

UO
ICUSTAY_ID       
200001.0    0.000
200003.0    0.320
200006.0    1.160
200007.0    0.620
200009.0    0.960
...           ...
299992.0    0.000
299993.0    0.245
299995.0    1.640
299998.0    0.730
299999.0    0.990

[53718 rows x 1 columns]

In [104]:
uo= uo.rename(columns = {'URINE_OUTPUT':'uo'})
uo

uo
ICUSTAY_ID       
4887256.0   2.106
5109404.0   3.085
4220812.0   2.835
4887256.0   2.365
1555036.0   1.495
...           ...
0.0         0.000
0.0         0.000
2885364.0   0.420
4808940.0   2.535
1202235.0   0.425

[37874 rows x 1 columns]

In [105]:
uo.loc[uo['uo'] <0.500, 'uo'] = 11
uo.loc[uo['uo'].between(0.500, 0.999), 'uo'] = 4
uo.loc[uo['uo']>0.999, 'uo'] = 0
uo

uo
ICUSTAY_ID     
4887256.0   0.0
5109404.0   0.0
4220812.0   0.0
4887256.0   0.0
1555036.0   0.0
...         ...
0.0         0.0
0.0         0.0
2885364.0   0.0
4808940.0   0.0
1202235.0   0.0

[37874 rows x 1 columns]

In [112]:
fields = ['ICUSTAY_ID', 'AGE']
age=pd.read_csv('icu.csv', usecols=fields, header=0, index_col=0)
age

AGE
ICUSTAY_ID            
280836       66.019178
206613       40.126027
220345       80.128767
249196       45.715068
210407       67.139726
...                ...
201233       77.161644
283653      300.200000
241585       53.128767
202802       67.884932
275346       34.241096

[61532 rows x 1 columns]

In [113]:
age['AGE'] = age['AGE'].apply(np.floor)
age

AGE
ICUSTAY_ID       
280836       66.0
206613       40.0
220345       80.0
249196       45.0
210407       67.0
...           ...
201233       77.0
283653      300.0
241585       53.0
202802       67.0
275346       34.0

[61532 rows x 1 columns]

In [114]:
age.loc[age['AGE'] <40, 'AGE'] = 0
age.loc[age['AGE'].between(40, 59), 'AGE'] = 7
age.loc[age['AGE'].between(60, 69), 'AGE'] = 12
age.loc[age['AGE'].between(70, 74), 'AGE'] = 15
age.loc[age['AGE'].between(75, 79), 'AGE'] = 16
age.loc[age['AGE']>79, 'AGE'] = 18

age

AGE
ICUSTAY_ID      
280836      12.0
206613       7.0
220345      18.0
249196       7.0
210407      12.0
...          ...
201233      16.0
283653      18.0
241585       7.0
202802      12.0
275346       0.0

[61532 rows x 1 columns]

In [115]:
age.to_csv('AGE.csv')

In [70]:

icu=pd.read_csv('icu.csv', header=0, index_col=0)
icu

SUBJECT_ID  HADM_ID  ICUSTAY_ID               INTIME  \
0             268   110404      280836  2198-02-14 23:27:38   
1             269   106296      206613  2170-11-05 11:05:29   
2             270   188028      220345  2128-06-24 15:05:20   
3             271   173727      249196  2120-08-07 23:12:42   
4             272   164716      210407  2186-12-25 21:08:04   
...           ...      ...         ...                  ...   
61527       94944   143774      201233  2104-04-15 10:18:16   
61528       94950   123750      283653  2155-12-08 05:33:16   
61529       94953   196881      241585  2160-03-03 16:09:11   
61530       94954   118475      202802  2183-03-25 09:53:10   
61531       94956   156386      275346  2157-05-19 02:54:54   

                   OUTTIME     LOS   ADMITTIME            DISCHTIME  \
0      2198-02-18 05:26:11  3.2490  2198-02-11  2198-02-18 03:55:00   
1      2170-11-08 17:46:57  3.2788  2170-11-05  2170-11-27 18:00:00   
2      2128-06-27 12:32:29  2.8939  2128-06-23  2128-06-27 12:31:00   
3      2120-08-10 00:39:04  2.0600  2120-08-07  2120-08-20 16:00:00   
4      2186-12-27 12:01:13  1.6202  2186-12-25  2187-01-02 14:57:00   
...                    ...     ...         ...                  ...   
61527  2104-04-17 14:51:00  2.1894  2104-04-11  2104-04-20 16:16:00   
61528  2155-12-10 17:24:58  2.4942  2155-12-07  2155-12-12 10:10:00   
61529  2160-03-04 14:22:33  0.9259  2160-03-03  2160-03-04 12:48:00   
61530  2183-03-27 17:55:03  2.3346  2183-03-25  2183-04-01 17:07:00   
61531  2157-05-23 14:58:04  4.5022  2157-05-19  2157-05-25 16:40:00   

      ADMISSION_TYPE  HOSPITAL_EXPIRE_FLAG GENDER         DOB  \
0          EMERGENCY                     1      F  2132-02-21   
1          EMERGENCY                     0      M  2130-09-30   
2           ELECTIVE                     0      M  2048-05-26   
3          EMERGENCY                     0      F  2074-11-30   
4          EMERGENCY                     0      M  2119-11-21   
...              ...                   ...    ...         ...   
61527      EMERGENCY                     0      M  2027-03-02   
61528      EMERGENCY                     0      F  1855-12-07   
61529       ELECTIVE                     0      F  2107-01-29   
61530       ELECTIVE                     0      F  2115-05-23   
61531      EMERGENCY                     0      M  2123-03-01   

                       DOD             DOD_HOSP  EXPIRE_FLAG         AGE  
0      2198-02-18 00:00:00  2198-02-18 00:00:00            1   66.019178  
1                      NaN                  NaN            0   40.126027  
2                      NaN                  NaN            0   80.128767  
3                      NaN                  NaN            0   45.715068  
4                      NaN                  NaN            0   67.139726  
...                    ...                  ...          ...         ...  
61527                  NaN                  NaN            0   77.161644  
61528                  NaN                  NaN            0  300.200000  
61529  2162-01-05 00:00:00                  NaN            1   53.128767  
61530                  NaN                  NaN            0   67.884932  
61531                  NaN                  NaN            0   34.241096  

[61532 rows x 16 columns]

In [71]:
icu=icu.drop(['DOD', 'DOD_HOSP', 'DOB', 'ADMITTIME', 'DISCHTIME', 'INTIME', 'OUTTIME'], axis=1)

In [75]:
icu.loc[icu['ADMISSION_TYPE'] == 'URGENT', 'uo'] = 8
icu.loc[icu['ADMISSION_TYPE'] == 'UEMERGENCY', 'uo'] = 8
icu.loc[icu['ADMISSION_TYPE'] == 'ELECTIVE', 'uo'] = 0
icu.loc[icu['ADMISSION_TYPE'] == 'NEWBORN', 'uo'] = 0

In [77]:
icu.loc[icu['AGE'] <40, 'AGE'] = 0
icu.loc[icu['AGE'].between(40, 59), 'AGE'] = 7
icu.loc[icu['AGE'].between(60, 69), 'AGE'] = 12
icu.loc[icu['AGE'].between(70, 74), 'AGE'] = 15
icu.loc[icu['AGE'].between(75, 79), 'AGE'] = 16
icu.loc[icu['AGE']>79, 'AGE'] = 18

In [79]:
icu = icu.rename(columns = {'LOS':'los','ADMISSION_TYPE':'admission', 
                              'HOSPITAL_EXPIRE_FLAG': 'hdeath', 'EXPIRE_FLAG':'death'})


In [66]:
fields = ['SUBJECT_ID', 'HADM_ID','UD_VALUE']
ud=pd.read_csv('ud_value.csv', usecols=fields, header=0, index_col=0)
ud

HADM_ID  UD_VALUE
SUBJECT_ID                   
109          172335         9
109          173633         9
109          131345         9
109          131376         9
109          135923         9
...             ...       ...
86684        158798        10
92644        132124        10
91832        173664        10
95280        155715         9
98701        124568         9

[6675 rows x 2 columns]

In [67]:
def adm_merge(table1, table2):
    return table1.merge(table2, how='left', left_on=['SUBJECT_ID','HADM_ID'], right_on=['SUBJECT_ID','HADM_ID'])

In [68]:
ud= ud.rename(columns = {'UD_VALUE':'ud'})

In [81]:
icu=adm_merge(icu, ud)
icu

SUBJECT_ID  HADM_ID  ICUSTAY_ID     los  admission  hdeath GENDER  \
0             268   110404      280836  3.2490  EMERGENCY       1      F   
1             269   106296      206613  3.2788  EMERGENCY       0      M   
2             270   188028      220345  2.8939   ELECTIVE       0      M   
3             271   173727      249196  2.0600  EMERGENCY       0      F   
4             272   164716      210407  1.6202  EMERGENCY       0      M   
...           ...      ...         ...     ...        ...     ...    ...   
61986       94944   143774      201233  2.1894  EMERGENCY       0      M   
61987       94950   123750      283653  2.4942  EMERGENCY       0      F   
61988       94953   196881      241585  0.9259   ELECTIVE       0      F   
61989       94954   118475      202802  2.3346   ELECTIVE       0      F   
61990       94956   156386      275346  4.5022  EMERGENCY       0      M   

       death   AGE   uo    ud  
0          1   0.0  NaN   NaN  
1          0   0.0  NaN  17.0  
2          0   0.0  0.0   NaN  
3          0   0.0  NaN   NaN  
4          0   0.0  NaN   NaN  
...      ...   ...  ...   ...  
61986      0   0.0  NaN   NaN  
61987      0  18.0  NaN   NaN  
61988      1   0.0  0.0   NaN  
61989      0   0.0  0.0   NaN  
61990      0   0.0  NaN   NaN  

[61991 rows x 11 columns]

In [82]:
icu.loc[icu['admission'] == 'URGENT', 'admission'] = 8
icu.loc[icu['admission'] == 'EMERGENCY', 'admission'] = 8
icu.loc[icu['admission'] == 'ELECTIVE', 'admission'] = 0
icu.loc[icu['admission'] == 'NEWBORN', 'admission'] = 0

In [84]:
icu=icu.drop(['uo', 'GENDER'], axis=1)

In [85]:
icu

SUBJECT_ID  HADM_ID  ICUSTAY_ID     los  admission  hdeath  death  \
0             268   110404      280836  3.2490          8       1      1   
1             269   106296      206613  3.2788          8       0      0   
2             270   188028      220345  2.8939          0       0      0   
3             271   173727      249196  2.0600          8       0      0   
4             272   164716      210407  1.6202          8       0      0   
...           ...      ...         ...     ...        ...     ...    ...   
61986       94944   143774      201233  2.1894          8       0      0   
61987       94950   123750      283653  2.4942          8       0      0   
61988       94953   196881      241585  0.9259          0       0      1   
61989       94954   118475      202802  2.3346          0       0      0   
61990       94956   156386      275346  4.5022          8       0      0   

        AGE    ud  
0       0.0   NaN  
1       0.0  17.0  
2       0.0   NaN  
3       0.0   NaN  
4       0.0   NaN  
...     ...   ...  
61986   0.0   NaN  
61987  18.0   NaN  
61988   0.0   NaN  
61989   0.0   NaN  
61990   0.0   NaN  

[61991 rows x 9 columns]

In [125]:
def icu_merge(table1, table2):
    return table1.merge(table2, how='left', left_on=['ICUSTAY_ID'], right_on=['ICUSTAY_ID'])

In [89]:
saps=icu_merge(icu, saps)

In [95]:
saps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61117 entries, 0 to 61116
Data columns (total 21 columns):
SUBJECT_ID     61117 non-null int64
HADM_ID        61117 non-null int64
ICUSTAY_ID     61117 non-null int64
los            61117 non-null float64
admission      61117 non-null int64
hdeath         61117 non-null int64
death          61117 non-null int64
age            61117 non-null float64
ud             61117 non-null float64
bun            52952 non-null float64
Bicarbonate    26060 non-null float64
ventilation    61117 non-null float64
Potassium      39731 non-null float64
Sodium         54983 non-null float64
Temp           52825 non-null float64
Bilirubin      24422 non-null float64
WBC            57479 non-null float64
hr             48225 non-null float64
gcs            52809 non-null float64
bp             60732 non-null float64
uo             189 non-null float64
dtypes: float64(15), int64(6)
memory usage: 10.3 MB


In [91]:
saps = saps.rename(columns = {'AGE':'age','BUN':'bun', 
                              'SystolicBP': 'bp', 'GCS':'gcs', 'HR':'hr'})

In [93]:
saps['ud'] = saps['ud'].fillna(0)

In [94]:
saps

SUBJECT_ID  HADM_ID  ICUSTAY_ID     los  admission  hdeath  death  \
0             268   110404      280836  3.2490          8       1      1   
1             269   106296      206613  3.2788          8       0      0   
2             270   188028      220345  2.8939          0       0      0   
3             271   173727      249196  2.0600          8       0      0   
4             272   164716      210407  1.6202          8       0      0   
...           ...      ...         ...     ...        ...     ...    ...   
61112       94944   143774      201233  2.1894          8       0      0   
61113       94950   123750      283653  2.4942          8       0      0   
61114       94953   196881      241585  0.9259          0       0      1   
61115       94954   118475      202802  2.3346          0       0      0   
61116       94956   156386      275346  4.5022          8       0      0   

        age    ud  bun  ...  ventilation  Potassium  Sodium  Temp  Bilirubin  \
0       0.0   0.0  6.0  ...          6.0        0.0     0.0   0.0        NaN   
1       0.0  17.0  0.0  ...          0.0        0.0     0.0   0.0        0.0   
2       0.0   0.0  0.0  ...          0.0        0.0     0.0   3.0        NaN   
3       0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        0.0   
4       0.0   0.0  0.0  ...          0.0        0.0     0.0   3.0        NaN   
...     ...   ...  ...  ...          ...        ...     ...   ...        ...   
61112   0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        NaN   
61113  18.0   0.0  0.0  ...          0.0        NaN     0.0   3.0        NaN   
61114   0.0   0.0  0.0  ...          6.0        NaN     0.0   3.0        NaN   
61115   0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        NaN   
61116   0.0   0.0  0.0  ...         11.0        0.0     0.0   3.0        NaN   

       WBC    hr   gcs    bp  uo  
0      0.0  11.0  26.0  13.0 NaN  
1      0.0   0.0   0.0   5.0 NaN  
2      0.0  11.0   0.0  13.0 NaN  
3      0.0   0.0   0.0   0.0 NaN  
4      0.0   0.0   0.0   5.0 NaN  
...    ...   ...   ...   ...  ..  
61112  0.0   2.0  26.0  13.0 NaN  
61113  0.0   2.0   5.0   0.0 NaN  
61114  0.0   NaN   7.0   5.0 NaN  
61115  0.0   NaN  26.0  13.0 NaN  
61116  0.0   NaN  26.0   5.0 NaN  

[61117 rows x 21 columns]

In [96]:
saps.to_csv('saps2_fnan.csv')

In [98]:
saps=saps.fillna(0)

In [100]:
saps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61117 entries, 0 to 61116
Data columns (total 21 columns):
SUBJECT_ID     61117 non-null int64
HADM_ID        61117 non-null int64
ICUSTAY_ID     61117 non-null int64
los            61117 non-null float64
admission      61117 non-null int64
hdeath         61117 non-null int64
death          61117 non-null int64
age            61117 non-null float64
ud             61117 non-null float64
bun            61117 non-null float64
Bicarbonate    61117 non-null float64
ventilation    61117 non-null float64
Potassium      61117 non-null float64
Sodium         61117 non-null float64
Temp           61117 non-null float64
Bilirubin      61117 non-null float64
WBC            61117 non-null float64
hr             61117 non-null float64
gcs            61117 non-null float64
bp             61117 non-null float64
uo             61117 non-null float64
dtypes: float64(15), int64(6)
memory usage: 10.3 MB


In [102]:
saps

SUBJECT_ID  HADM_ID  ICUSTAY_ID     los  admission  hdeath  death  \
0             268   110404      280836  3.2490          8       1      1   
1             269   106296      206613  3.2788          8       0      0   
2             270   188028      220345  2.8939          0       0      0   
3             271   173727      249196  2.0600          8       0      0   
4             272   164716      210407  1.6202          8       0      0   
...           ...      ...         ...     ...        ...     ...    ...   
61112       94944   143774      201233  2.1894          8       0      0   
61113       94950   123750      283653  2.4942          8       0      0   
61114       94953   196881      241585  0.9259          0       0      1   
61115       94954   118475      202802  2.3346          0       0      0   
61116       94956   156386      275346  4.5022          8       0      0   

        age    ud  bun  ...  ventilation  Potassium  Sodium  Temp  Bilirubin  \
0       0.0   0.0  6.0  ...          6.0        0.0     0.0   0.0        0.0   
1       0.0  17.0  0.0  ...          0.0        0.0     0.0   0.0        0.0   
2       0.0   0.0  0.0  ...          0.0        0.0     0.0   3.0        0.0   
3       0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        0.0   
4       0.0   0.0  0.0  ...          0.0        0.0     0.0   3.0        0.0   
...     ...   ...  ...  ...          ...        ...     ...   ...        ...   
61112   0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        0.0   
61113  18.0   0.0  0.0  ...          0.0        0.0     0.0   3.0        0.0   
61114   0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        0.0   
61115   0.0   0.0  0.0  ...          6.0        0.0     0.0   3.0        0.0   
61116   0.0   0.0  0.0  ...         11.0        0.0     0.0   3.0        0.0   

       WBC    hr   gcs    bp   uo  
0      0.0  11.0  26.0  13.0  0.0  
1      0.0   0.0   0.0   5.0  0.0  
2      0.0  11.0   0.0  13.0  0.0  
3      0.0   0.0   0.0   0.0  0.0  
4      0.0   0.0   0.0   5.0  0.0  
...    ...   ...   ...   ...  ...  
61112  0.0   2.0  26.0  13.0  0.0  
61113  0.0   2.0   5.0   0.0  0.0  
61114  0.0   0.0   7.0   5.0  0.0  
61115  0.0   0.0  26.0  13.0  0.0  
61116  0.0   0.0  26.0   5.0  0.0  

[61117 rows x 21 columns]

In [117]:
saps=saps.drop(['age'], axis=1)

In [118]:
saps=icu_merge(saps, age)

In [119]:
saps

SUBJECT_ID  HADM_ID  ICUSTAY_ID     los  admission  hdeath  death  \
0             268   110404      280836  3.2490          8       1      1   
1             269   106296      206613  3.2788          8       0      0   
2             270   188028      220345  2.8939          0       0      0   
3             271   173727      249196  2.0600          8       0      0   
4             272   164716      210407  1.6202          8       0      0   
...           ...      ...         ...     ...        ...     ...    ...   
61112       94944   143774      201233  2.1894          8       0      0   
61113       94950   123750      283653  2.4942          8       0      0   
61114       94953   196881      241585  0.9259          0       0      1   
61115       94954   118475      202802  2.3346          0       0      0   
61116       94956   156386      275346  4.5022          8       0      0   

         ud  bun  Bicarbonate  ...  Potassium  Sodium  Temp  Bilirubin  WBC  \
0       0.0  6.0          0.0  ...        0.0     0.0   0.0        0.0  0.0   
1      17.0  0.0          0.0  ...        0.0     0.0   0.0        0.0  0.0   
2       0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
3       0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
4       0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
...     ...  ...          ...  ...        ...     ...   ...        ...  ...   
61112   0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
61113   0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
61114   0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
61115   0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   
61116   0.0  0.0          0.0  ...        0.0     0.0   3.0        0.0  0.0   

         hr   gcs    bp   uo   AGE  
0      11.0  26.0  13.0  0.0  12.0  
1       0.0   0.0   5.0  0.0   7.0  
2      11.0   0.0  13.0  0.0  18.0  
3       0.0   0.0   0.0  0.0   7.0  
4       0.0   0.0   5.0  0.0  12.0  
...     ...   ...   ...  ...   ...  
61112   2.0  26.0  13.0  0.0  16.0  
61113   2.0   5.0   0.0  0.0  18.0  
61114   0.0   7.0   5.0  0.0   7.0  
61115   0.0  26.0  13.0  0.0  12.0  
61116   0.0  26.0   5.0  0.0   0.0  

[61117 rows x 21 columns]

In [130]:
saps.to_csv('saps2_f.csv')

In [122]:
fields = ['ICUSTAY_ID', 'UO']
uo=pd.read_csv('uo_value.csv', usecols=fields, header=0, index_col=0)
uo

UO
ICUSTAY_ID       
200001.0    0.000
200003.0    0.320
200006.0    1.160
200007.0    0.620
200009.0    0.960
...           ...
299992.0    0.000
299993.0    0.245
299995.0    1.640
299998.0    0.730
299999.0    0.990

[53718 rows x 1 columns]

In [123]:
saps=saps.drop(['uo'], axis=1)

In [127]:
saps=icu_merge(saps, uo)

In [128]:
saps=saps.fillna(0)

In [129]:
saps.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61117 entries, 0 to 61116
Data columns (total 21 columns):
SUBJECT_ID     61117 non-null int64
HADM_ID        61117 non-null int64
ICUSTAY_ID     61117 non-null int64
los            61117 non-null float64
admission      61117 non-null int64
hdeath         61117 non-null int64
death          61117 non-null int64
ud             61117 non-null float64
bun            61117 non-null float64
Bicarbonate    61117 non-null float64
ventilation    61117 non-null float64
Potassium      61117 non-null float64
Sodium         61117 non-null float64
Temp           61117 non-null float64
Bilirubin      61117 non-null float64
WBC            61117 non-null float64
hr             61117 non-null float64
gcs            61117 non-null float64
bp             61117 non-null float64
AGE            61117 non-null float64
UO             61117 non-null float64
dtypes: float64(15), int64(6)
memory usage: 10.3 MB


In [131]:
saps_nan=pd.read_csv('saps2_fnan.csv', header=0, index_col=0)

In [132]:
saps_nan=icu_merge(saps_nan, uo)

In [134]:
saps_nan=icu_merge(saps_nan, age)

In [135]:
saps_nan.to_csv('saps2_nan.csv')

## Data Wrangling Summary:
+ death - is a binary flag which indicates whether the patient died, i.e.whether DOD is null or not. These deaths include both deaths within the hospital ( DOD_HOSP ) and deaths identified by matching the patient to the social security master death index ( DOD_SSN ).
+ The final 38 columns(variables) were selected and created from a total of 324 columns found in 26 relational tabeles of MIMIC-IIIv1.4 database.
+ We have 46,234 unique patients in the final clinical deterioration dataset. Around 286 were lost during data cleaning. MIMIC-IIIv1.4 database has a total population of 46,520 patients.
+ There are 60,517 unique Intensive Care Unit(ICU) stays, some patients were addmitted to ICU more than once. The entire MIMIC-IIIv1.4 datasets has 61,532 ICU stays. Only one ICU stay per hospital amission were included in the study, as a result 1,015 ICU stays were excluded. 
+ There are 800,097 electronic chart events from 4,602 patients. The primary repository of a patients's information is their electronic chart. Each raw in this the final 'Clinical Deterioration' dataset represent a unique entry of patient electronic chart. 
+ The data spans June 2001 - October 2011. In this dataset, the years were moved to future years to comply with HIPAA.